### PR TITLE
Credits ledger

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description

This PR adds a "ledger" (a chronological list of all purchases made with credits) to the credits page.

I've introduced a `Credits::Buyer` class to purchase things (for now only listings):

```ruby
Credits::Buyer.call(
  purchaser: user_or_org,
  purchase: listing,
  cost: 10,
)
```

and a `Credits::Ledger` to list the items.

The `credits` table loses `spent_on`, gains `purchase_id` and `purchase_type` so that each credit is attached to a purchase and gains a `spent_at` to track when the purchase was made.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

Screenshot_2019-07-08 Credits - DEV(local) Community 👩‍💻👨‍💻

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/controllers/classified_listings_controller.rb [hunk 1]
```diff
@@ -9,69 +9,63 @@ def index
...
     @organizations = current_user.organizations
-    @credits = current_user.credits.where(spent: false)
+    @credits = current_user.credits.unspent
   end
 
   def edit
     authorize @classified_listing
     @organizations = current_user.organizations
-    @credits = current_user.credits.where(spent: false)
+    @credits = current_user.credits.spent
```
Comment (lightalloy) [inline app/controllers/classified_listings_controller.rb]: Is it intended that the `spent` credits are queued instead of `unspent`?

Comment (rhymes) [reply app/controllers/classified_listings_controller.rb]: No, definitely an error, thanks!

File: app/controllers/classified_listings_controller.rb [hunk 2]
```diff
@@ -96,6 +92,32 @@ def dashboard
...
   private
 
+  def create_listing(purchaser, cost)
+    # this will 500 for now if they don't belong in the org
+    authorize @classified_listing, :authorized_organization_poster? if @classified_listing.organization_id.present?
+
+    @classified_listing.bumped_at = Time.current
+    @classified_listing.published = true
+
+    if @classified_listing.save
```
Comment (lightalloy) [inline app/controllers/classified_listings_controller.rb]: It seems like bumping a listing (saving) and spending credits (marking as spent) should be inside a transaction. What do you think about it?

Comment (rhymes) [reply app/controllers/classified_listings_controller.rb]: It's a good call, there shouldn't be a situation where the listing is on the DB but the credits haven't been subtracted. Same with all the other operations that spend credits.
Let me see what's the best way to go about that without revolutionizing the code

File: app/services/credits/ledger.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    def build_ledger_for(credits)
+      purchases = load_purchases(credits)
+      items = []
+
+      # to avoid N+1 on purchases, we load them by type separately
+      listings_purchases = purchases.select { |row| row.purchase_type == "ClassifiedListing" }
+      listings = ClassifiedListing.where(id: listings_purchases.map(&:purchase_id))
+      listings_purchases.each do |purchase|
+        listing = listings.select { |l| l.id == purchase.purchase_id }.first
```
Comment (lightalloy) [inline app/services/credits/ledger.rb]: `select` + `first` can be replaced with `detect`

File: spec/services/credits/buyer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  context "when enough credits are available" do
+    it "spends credits for the user" do
+      create_list(:credit, 2, user: user)
+      expect do
+        res = described_class.call(purchaser: user, purchase: listing, cost: 2)
+        expect(res).to be(true)
+      end.to change(user.credits.spent, :count)
+    end
+
+    it "does not spend credits for the organization" do
```
Comment (lightalloy) [inline spec/services/credits/buyer_spec.rb]: The description says it does NOT spend, but the test asserts that it spends actually?

Comment (rhymes) [reply spec/services/credits/buyer_spec.rb]: yeah good catch, I copied the description from the previous block, I'll fix it right away

File: app/services/credits/ledger.rb (hunk 3 above)
Comment (rhymes) [reply app/services/credits/ledger.rb]: don't know why Rubocop didn't catch this hahaha

**Pull request merged**